framework/LearningGate.py

```diff
@@ -173,6 +217,393 @@ def train(self,trainingSet):
+#        # weight is variance: MORE variance means MORE importance
+#        std = np.std(clusterFeatureDict[feat])
+#        weight = std
+#      else:
+#        # groupWeight = 1.0 / float(len(featureGroups))
+#        # weight = groupWeight / float(len(featureGroups[metric]))
+#        # normalize weights later
+#        weight = 1.0 / float(len(featureGroups[metric]))
+#      # DEBUGG
+#      # apply special weighting
+#      if metric == 'Basic' and ID in ['mean','min','max']:
+#        weight *= 2
+#      # scale training points by weights
+#      # TODO do this after normalization # clusterFeatureDict[feat] *= weight
+#      weights[f] = weight
+#    # normalize weights
+#    ## METHOD: sum of weights should be unity
+#    scale = np.sum(weights)
+#    ## METHOD: by volume, assuming all weights are 1.0 initially before preference
+#    # vol = np.product(list(np.max(v) for v in clusterFeatureDict.values()))
+#    # print('DEBUGG original volume:',vol)
+#    # renormalize the entirety of the space to have the same hypervolume as before weighting
+#    # newVolume = np.product(weights)
+#    # oldVolume = 1.0 # because we scaled between 0 and 1, this will fail if you don't
+#    # scale = (oldVolume/newVolume)**(1.0/float(len(features)))
+#    ## END by volume
+#    for feature,vals in clusterFeatureDict.items():
+#      clusterFeatureDict[feature] = vals * scale
+#      v = clusterFeatureDict[feature]
+#      print('DEBUGG val range: {:15.15s} {:1.3e} {:1.3e} {:1.3e}'.format(feature,np.min(v),np.average(v),np.max(v)))
+#    vol = np.product(list(np.max(v) for v in clusterFeatureDict.values()))
+#    print('DEBUGG volume:',vol)
+#    return clusterFeatureDict
+
+  def _trainSubdomainRoms(self, templateRom, counter, trainingSet, strategy):
+    """
+      Trains the ROMs on each clusterable subdomain, and calculates features based on the data, rom
+      @ In, templateRom, SupervisedLearning instance, base ROM as a template for training
+      @ In, counter, list(tuple), instructions for subdividing domain into subdomains
+      @ In, trainingSet, dict, data on which ROMs should be trained
+      @ In, strategy, str, clustering strategy (e.g. "segment", "continuous", "cluster")
+      @ Out, clusterFeatureDict, dict, clustering information as {feature: [rom values]}
+      @ Out, roms, np.array(SubpervisedLearning instances), trained ROMs for each subdomain
+    """
+    targets = templateRom.target[:]
+    targets.remove(templateRom.pivotParameterID)
+    clusterFeatureDict = {}
+    roms = []
+    if self.targetDatas is None:
+      self.targetDatas = [] # DEBUGG only
+    for i,subdiv in enumerate(counter):
+      # slicer for data selection
+      picker = slice(subdiv[0], subdiv[-1]+1)
+      ## TODO only consider one sample at 0? Should do more for non-ARMA ROMs!
+      ##  -> for now, only ARMAs can be used with this method, so we can address this later
+      data = dict((var,[trainingSet[var][0][picker]]) for var in trainingSet)
```

Comment:
This might be true. Given a ROM with several indices, it may be that not all should be subspaces, and not treated in the same way. We should definitely do some designing when we want to move to introduce ROMs with large numbers of samples into this methodology.